jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -825,314 +439,372 @@ private boolean wakeup()
         {
-            return _content;
+            return _rawContentArrived;
         }
 
-        @Override
-        public InvocationType getInvocationType()
+        boolean hasRawContent()
         {
-            return InvocationType.NON_BLOCKING;
+            return _rawContent != null;
         }
 
-        public int get(byte[] buffer, int offset, int length)
+        Interceptor getInterceptor()
         {
-            length = Math.min(_content.remaining(), length);
-            _content.get(buffer, offset, length);
-            return length;
+            return _interceptor;
         }
 
-        public int skip(int length)
+        void setInterceptor(Interceptor interceptor)
         {
-            length = Math.min(_content.remaining(), length);
-            _content.position(_content.position() + length);
-            return length;
+            this._interceptor = interceptor;
         }
 
-        public boolean hasContent()
+        void addContent(Content content)
         {
-            return _content.hasRemaining();
-        }
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} addContent {}", this, content);
+            if (content == null)
+                throw new AssertionError("Cannot add null content");
+            if (_consumeFailure != null)
+            {
+                content.failed(_consumeFailure);
+                return;
+            }
+            if (_rawContent != null)
+                throw new AssertionError("Cannot add new content while current one hasn't been processed");
 
-        public int remaining()
-        {
-            return _content.remaining();
+            _rawContent = content;
+            _rawContentArrived += content.remaining();
         }
 
-        public boolean isEmpty()
+        void consumeTransformedContent(Consumer<Throwable> failRawContent, Throwable failure)
         {
-            return !_content.hasRemaining();
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} consumeTransformedContent", this);
+            // start by depleting the current _transformedContent
+            if (_transformedContent != null)
+            {
+                _transformedContent.skip(_transformedContent.remaining());
+                if (_transformedContent != _rawContent)
+                    _transformedContent.succeeded();
+                _transformedContent = null;
+            }
+
+            // don't bother transforming content, directly deplete the raw one
+            if (_rawContent != null)
+            {
+                _rawContent.skip(_rawContent.remaining());
+                _rawContent.succeeded();
+                _rawContent = null;
+            }
+
+            // fail whatever other content the producer may have
+            _consumeFailure = failure;
+            failRawContent.accept(failure);
         }
 
-        @Override
-        public String toString()
+        int available(Runnable rawContentProducer)
```

Comment:
Does this still need the rawContentProducer passed in?

Reply:
It's that or making `ContentProducer` non-static. I personally prefer keeping it static but I'm not married to that idea.